src/sinks/honeycomb.rs

```diff
@@ -0,0 +1,143 @@
+use crate::{
+    dns::Resolver,
+    event::{log_schema, Event, Value},
+    sinks::util::http::{https_client, BatchedHttpSink, HttpSink},
+    sinks::util::{BatchBytesConfig, BoxedRawValue, JsonArrayBuffer, TowerRequestConfig, UriSerde},
+    tls::TlsSettings,
+    topology::config::{DataType, SinkConfig, SinkContext, SinkDescription},
+};
+use futures::{compat::Future01CompatExt, TryFutureExt};
+use futures01::Stream;
+use http::{Request, StatusCode, Uri};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+lazy_static::lazy_static! {
+    static ref HOST: UriSerde = Uri::from_static("https://api.honeycomb.io/1/batch").into();
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct HoneycombConfig {
+    api_key: String,
+
+    // TODO: we probably want to make this a template
+    // but this limits us in how we can do our healthcheck.
+    dataset: String,
+
+    #[serde(default)]
+    batch: BatchBytesConfig,
+
+    #[serde(default)]
+    request: TowerRequestConfig,
+}
+
+inventory::submit! {
+    SinkDescription::new_without_default::<HoneycombConfig>("honeycomb")
+}
+
+#[typetag::serde(name = "honeycomb")]
+impl SinkConfig for HoneycombConfig {
+    fn build(&self, cx: SinkContext) -> crate::Result<(super::RouterSink, super::Healthcheck)> {
+        let request_settings = self.request.unwrap_with(&TowerRequestConfig::default());
+        let batch_settings = self.batch.unwrap_or(bytesize::mib(5u64), 1);
+
+        let sink = BatchedHttpSink::new(
+            self.clone(),
+            JsonArrayBuffer::default(),
+            request_settings,
+            batch_settings,
+            None,
+            &cx,
+        );
+
+        let healthcheck = Box::new(Box::pin(healthcheck(self.clone(), cx.resolver())).compat());
+
+        Ok((Box::new(sink), healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn sink_type(&self) -> &'static str {
+        "honeycomb"
+    }
+}
+
+impl HttpSink for HoneycombConfig {
+    type Input = serde_json::Value;
+    type Output = Vec<BoxedRawValue>;
+
+    fn encode_event(&self, event: Event) -> Option<Self::Input> {
+        let mut log = event.into_log();
+
+        let timestamp = if let Some(Value::Timestamp(ts)) = log.remove(log_schema().timestamp_key())
+        {
+            ts
+        } else {
+            chrono::Utc::now()
+        };
+
+        Some(json!({
+            "timestamp": timestamp.to_rfc3339_opts(chrono::SecondsFormat::Nanos, true),
+            "data": log.all_fields(),
+        }))
+    }
+
+    fn build_request(&self, events: Self::Output) -> http::Request<Vec<u8>> {
+        let uri = self.build_uri();
+        let mut request = Request::post(uri);
+
+        request.header("X-Honeycomb-Team", self.api_key.clone());
```

Comment:
Liz from the honeycomb asked us to set a unique user agent here.

Reply:
A unique vector user agent like this issue? https://github.com/timberio/vector/issues/501

Reply:
Exactly.